### ISSUE TITLE
Fix timestamps

### ISSUE THREAD

**Issue by larsbrinkhoff** [PDP-10/its#2051]
Update the file build/timestamps.txt and be careful to ensure they are correct.

Also look into using a version of touch or gtouch that handles symlinks well.

**Comment by larsbrinkhoff** [comment]
See comments here: https://github.com/PDP-10/its/issues/2044#issuecomment-870858093

**Comment by cstacy** [comment]
Please see this file 2016-0702-touch.txt (comments and testing).

[2016-0702-touch.txt](https://github.com/PDP-10/its/files/6757445/2016-0702-touch.txt)


Also attached is a build illustrating the issue.

[2021-0702-build.txt](https://github.com/PDP-10/its/files/6757447/2021-0702-build.txt)

**Comment by larsbrinkhoff** [comment]
Thanks, fixing touch to work right with softlinks is also part of this.  Added to the top description.

**Comment by larsbrinkhoff** [comment]
@cstacy, you ask:

```
QUESTIONS:

* Are some people NOT getting the error messages?

* I'm not sure what is intended.
  Are there supposed to be files there at this point?
  Or is it maybe trying to create artifacts that are
  to go onto a virtual tape? With the -h flag it seems
  that symlink are supposed to be involved...
```

I don't get the error messages running on Ubuntu 16:

```
Synchronizing submodule url for 'tools/pdp6'
git submodule update --recursive --init `dirname tools/pdp6/.gitignore`
build/stamp.sh build/timestamps.txt
mkdir -p out/klh10/stamp
touch out/klh10/stamp/touch
mkdir -p out/klh10/system
```

The intent is that `touch -h` should create a symlink and set its timestamp.  Maybe the touch you ran doesn't create symlinks?  If so, the script could be updated to first make the symlink, then use touch -h to set its timestamp.

**Comment by larsbrinkhoff** [comment]
@cstacy and @eswenson1, what happens on a Mac for various permutations of touch, gtouch, /usr/bin/touch, etc using -h and a filename that either doesn't exist, or exists and is a symlink?

**Comment by cstacy** [comment]
See my notes!


:)

**Comment by cstacy** [comment]
"touch -h" on all platforms is documented NOT to ever create any kind of file. It does not create symbolic links. This is POSIX.

In addition to the test in my notes above, I also tested it just now on a Linux machine. It works correctly there, too: no file is created.  It gives the "No such file" error message.

The only difference between platforms is that on some of them, it fails silently rather than giving the error message.

If this script was trying to create symbolic links this way, I submit that it has never worked on any platform. But maybe nobody noticed because the command will fail silently.

If you see symbolic links being created, it's not from this command.

(Linux 4.9, GNU Coreutils 8.26)

**Comment by cstacy** [comment]
Oh, there is another difference between platforms. When touch UPDATES an existing symbolic link, on macOS (aka BSD), it accepts the invalid time stamps and adjusts them to a different value.

But it's not what is creating the files.
Not with "-h" it's not.

**Comment by cstacy** [comment]
I think the files must already be created by the time the timestamps.sh script is run. (I presume they must be symbolic links, because otherwise, why the "-h" flag.)

Therefore, if  macOS is the only platform getting the "file not found" error message, perhaps that's the only platform that doesn't have the files cloned when the script runs.

The touch command on Linux gives "file not found" messages.
But the script is not giving them.
(The script is giving the "invalid date format" messages.)

It would seem that the macOS version is missing files that exist on  the Linux version. And those files are NOT being created by the touch command in timestamps.h -- I tested that.


Maybe a submodule is missing in the macOS build?

[2021-0704-touch.txt](https://github.com/PDP-10/its/files/6759468/2021-0704-touch.txt)

**Comment by larsbrinkhoff** [comment]
Oh sorry, my mistake.  I see now the symlinks do exist before touch is run.  So it's just supposed to set their timestamps.  The -h is there to affect the links rather than their targets (which don't exist because they're only meaningful to itstar).

**Comment by cstacy** [comment]
I am boggled.

[2021-0704-trial.txt](https://github.com/PDP-10/its/files/6759585/2021-0704-trial.txt)

**Comment by larsbrinkhoff** [comment]
I think in some circumstances "file not found" is due to the symlink pointing to a file which doesn't exist.  Note that `bin/_teco_/ts.rmode` points to `_teco_/mod929.>` but there is no file `bin/_teco_/_teco_/mod929.>`.  There *is* a file `bin/_teco_/mod928.177114` but there's not where the symlink is pointing.

You tested running touch from bash.  Does that use GNU touch?

**Comment by larsbrinkhoff** [comment]
Of course "touch -h" should just update the symlink timestamp and presumably not care what it points to.  But from your log it seems it does care, maybe?  I'm not sure how many distinct versions of touch you have between /usr/bin, homebrew, gtouch etc, but maybe try them all with regards to dangling symlinks?

**Comment by larsbrinkhoff** [comment]
At some point I considered adding a small home made "touch.c" that did the right thing with symlinks, but then it seems all platforms that has the lutimes system call also has "touch -h".  But if some implementations of "touch -h" are buggy maybe that's the way forward.

**Comment by cstacy** [comment]
Some stats:

There are 3165 touch commands.
There are 66 failures.
There are 70 symbolic links.

Not every symbolic link fails.
I see no pattern to the link or target names.

**Comment by cstacy** [comment]
Haha, one thing that crossed my mind early on was, "Could this be using some local version of touch that is in the build?"

Maybe that's the solution, but is it really possible that touch is broken? BOTH VERSIONS -- GNU and BSD?

Seems more likely that either somehow (a) Despite all evidence to the contrary (different shells, explicit fully qualified program names on the command line, etc.) that I am not really running the "touch" that I think I am?  or   (b) the system call is broken in macOS. The later would be harder to believe.

But maybe we need "itstouch" to go with "itstar".
I would start by copying the source code of GNU touch and giving it a different name, and have the script call it by that different name.

But I sure would like to know what's going on.

I'm going to call a friend of mine who is a Unix sysadmin wizard and also a Mac fan, and see if he has any thoughts.

**Comment by cstacy** [comment]
I know not what devil has possessed my "touch" or quite how or why it is screwing me.

But copying /usr/local/bin/gtouch (Homebrew) to "itstouch", and modifying the stamp.sh script to call it, solves the problem.

Now all I have is the invalid date problem, which will be fixed by getting rid of the bad dates in the source, I guess.

**Comment by larsbrinkhoff** [comment]
Great, thanks!

**Comment by drboone** [comment]
Show me on the doll where the bad OS `touch`ed you? :)

**Comment by cstacy** [comment]
Now magically, my touch problems have gone away without the hacks.

My theory is now that:  (1) in the default environment, I had a bad version of touch that was giving the file-not-found errors. What version of touch is even allowed to EVER say that? (2) When I switching things around by installing GNU touch, and was testing that, SOMEHOW I was not running what I thought. This would suggest that the "which" command was lying to me, and the shell had cached the bad version.

This does not explain transcripts where I explicitly invoked full-pathname versions of touch and still lost. And under different shells. I have not changed any init files or anything, either. But now all the sudden shit is working again.

What I did  do, is put  my touch->gtouch link back and reboot.  But that situation already existed under lossage scenarios: I was rebooting across those many experiments, too.

So, I still have no idea what the nightmare was .

But today I woke up?

**Comment by cstacy** [comment]
Reminder: zsh on macOS has some weird built-in "touch", which maybe is that script I found in some zsh library of some sort that looked like it was trying to DWIM-ify touch commands. Maybe it was having some filename handling bug or other stupidity.  At no time does the (also built-in, I believe) "which" command report "builtin" as you would expect. It always reports what you would expect based on $PATH.

That doesn't quite explain things, but I'm just reiterating it, as it is the most suspicious thing I came across when investigating the failure.

**Comment by eswenson1** [comment]
That's an interesting find, Chris. I use zsh on my Mac too.

**Comment by eswenson1** [comment]
I just tried a klh10 build using the current (as of now) master branch.  I got these three timestamp-related warnings:

```
touch: invalid date format ‘198904020211.39’
touch: invalid date format ‘198904020211.45’
touch: invalid date format ‘198904020211.46’
```
I'm running Ubuntu 18.04 LTS on this machine.

**Comment by larsbrinkhoff** [comment]
This should now be fixed by #2074.
